rpcs3/rpcs3qt/log_frame.cpp

```diff
@@ -430,41 +430,41 @@ void log_frame::UpdateUI()
 {
 	const auto start = steady_clock::now();
 
+	const auto tty = g_fxo->get<tty_t>();
+
 	// Check TTY logs
-	while (const u64 size = std::max<s64>(0, g_tty_size.load() - (m_tty_file ? m_tty_file.pos() : 0)))
+	if (tty && tty->fd && m_tty_time != tty->time)
 	{
-		std::string buf;
-		buf.resize(size);
-		buf.resize(m_tty_file.read(&buf.front(), buf.size()));
+		// save old selection
+		QTextCursor text_cursor{ m_tty->document() };
+		const int sel_pos = text_cursor.position();
+		int sel_start = text_cursor.selectionStart();
+		int sel_end = text_cursor.selectionEnd();
+
+		m_tty->clear();
+		m_tty_time = tty->time;
 
-		if (buf.find_first_of('\0') != umax)
+		std::string buf;
 		{
-			m_tty_file.seek(s64{0} - buf.size(), fs::seek_mode::seek_cur);
-			break;
+			std::lock_guard lock(tty->mtx);
+			const auto pos = tty->fd.pos();
+			buf = tty->fd.to_string();
+			tty->fd.seek(pos);
 		}
 
-		if (!buf.empty() && m_TTYAct->isChecked())
+		if (m_TTYAct->isChecked())
 		{
-			std::stringstream buf_stream;
-			buf_stream.str(buf);
-			std::string buf_line;
-			while (std::getline(buf_stream, buf_line))
+			for (usz i = 0, next = buf.find_first_of('\n'); i < buf.size(); i = std::exchange(next, std::min<usz>(buf.size(), buf.find_first_of('\n', next + 1))) + 1)
```

Comment:
there is zero comment for this complicated line of code

Reply:
I added some.